modules/consensus/setdb.go

```diff
@@ -0,0 +1,226 @@
+		siacoinOutputDiffs:        pb.SiacoinOutputDiffs,
+		fileContractDiffs:         pb.FileContractDiffs,
+		siafundOutputDiffs:        pb.SiafundOutputDiffs,
+		delayedSiacoinOutputDiffs: pb.DelayedSiacoinOutputDiffs,
+		siafundPoolDiffs:          pb.SiafundPoolDiffs,
+
+		consensusSetHash: pb.ConsensusSetHash,
+	}
+	return bn
+}
+
+func openDB(filename string) (*setDB, error) {
+	db, err := persist.OpenDatabase(meta, filename)
+	if err != nil {
+		return nil, err
+	}
+
+	var buckets []string = []string{
+		"Path", "BlockMap",
+	}
+
+	// Create buckets
+	err = db.Update(func(tx *bolt.Tx) error {
+		for _, bucketName := range buckets {
+			_, err := tx.CreateBucketIfNotExists([]byte(bucketName))
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return &setDB{db}, nil
+}
+
+// addItem and getItem are part of consensus due to stricter error conditions.
+//
+// addItem should only be called from this file, and adds a new item
+// to the database
+func (db *setDB) addItem(bucket string, key, value interface{}) error {
+	v := encoding.Marshal(value)
+	k := encoding.Marshal(key)
+	return db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucket))
+		if build.DEBUG {
+			if b == nil {
+				panic(errNilBucket)
+			}
+			i := b.Get(k)
+			if i != nil {
+				panic(errBadSetInsert)
+			}
+		}
+		return b.Put(k, v)
+	})
+}
+
+func (db *setDB) getItem(bucket string, key interface{}) (item []byte, err error) {
+	k := encoding.Marshal(key)
+	err = db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucket))
+		if build.DEBUG {
+			if b == nil {
+				panic(errNilBucket)
+			}
+		}
+		item = b.Get(k)
+		if build.DEBUG {
+			if item == nil {
+				panic(errNilItem)
+			}
+		}
+		return nil
+	})
+	return
+}
+
+// AddBlock inserts a block into the database at the "end" of the chain, i.e.
+// the current height + 1.
+func (db *setDB) addPath(block types.Block) error {
+	value := encoding.Marshal(block.ID())
+	return db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte("Path"))
+		key := encoding.EncUint64(uint64(b.Stats().KeyN))
+		return b.Put(key, value)
+	})
+}
+
+// RemoveBlock removes a block from the "end" of the chain, i.e. the block
+// with the largest height.
+func (db *setDB) rmPath() error {
+	return db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte("Path"))
+		key := encoding.EncUint64(uint64(b.Stats().KeyN - 1))
+		return b.Delete(key)
+	})
+}
+
+// path retreives the block id of a block at a given hegiht from the path
+func (db *setDB) path(h types.BlockHeight) (id types.BlockID, err error) {
+	idBytes, err := db.getItem("Path", h)
+	if err != nil {
+		return
+	}
+	err = encoding.Unmarshal(idBytes, &id)
+	return
+}
+
+// pathHeight returns the size of the current path
+func (db *setDB) pathHeight() (types.BlockHeight, error) {
+	h, err := db.BucketSize("Path")
+	return types.BlockHeight(h), err
+}
+
+// addBlockMap adds a block node to the block map
+func (db *setDB) addBlockMap(bn blockNode) error {
```

Comment:
hmm.

I don't think that it's ambigious, because you are calling db.AddProcessedBlock. The consensus package distinguishes between `add` and `commit` as well.